### PR TITLE
Scripts - Add Switch functionality

### PR DESCRIPTION
**Pull request by TyroneMF**
- Adds switch function, Adds ACE interaction and capability to switch controlling power levels.

### CODE REVIEW

File: Scripts/fn_switch.sqf [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    _args params ["_object", "_powerOne", "_powerTwo", ["_switchLastState", -1]];
+
+    // Check all animation sources on objects to see if switched or not. (1: ON / -1: OFF)
+    private _switch = _object animationSourcePhase "SwitchPosition";
+
+    // Update last state
+    _args set [3, _switch];
+
+    if (_switch != _switchLastState) then {
+        if (_switch == 1) then {
```
Comment (jonpas) [inline Scripts/fn_switch.sqf:40]:
```suggestion
        if (_switch == 1) exitWith {
```
To make it clearer that next `_switch == -1` will not ever run.

File: Scripts/fn_switch.sqf [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            _object animateSource ["Power_1", _powerOne, 0.25];
+            _object animateSource ["Power_2", _powerTwo, 0.25];
+            _object setObjectTextureGlobal [1, "#(argb,8,8,3)color(0,1,0,0.05,ca)"];
+            _object setVariable ["TAC_Switched", true, true];
+        };
+        if (_switch == -1) then {
```
Comment (jonpas) [inline Scripts/fn_switch.sqf:46]:
```suggestion
        if (_switch == -1) exitWith {
```

File: Scripts/fn_switch.sqf [hunk 3]
```diff
@@ -0,0 +1,53 @@
+#include "..\script_component.hpp"
+
+/*
+ * Author: Tyrone
+ * Functionality for toggling switches, When "OFF" switch is red and power levels are 0, when "ON" switch is Green and power levels are either 1 or player defined.
+ *
+ * Requires fn_switchAction.sqf
```
Comment (jonpas) [inline Scripts/fn_switch.sqf:7]: Why? This function could easily add actions as well, you already have require the object as argument.

File: Scripts/fn_switch.sqf [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            _object setVariable ["TAC_Switched", true, true];
+        };
+        if (_switch == -1) then {
+            _object animateSource ["Power_1", 0, 0.25];
+            _object animateSource ["Power_2", 0, 0.25];
+            _object setObjectTextureGlobal [1, "#(argb,8,8,3)color(1,0,0,0.05,ca)"];
+            _object setVariable ["TAC_Switched", false, true];
+        };
+    };
+}, 1, [_object, _powerOne, _powerTwo]] call CBA_fnc_addPerFrameHandler;
```
Comment (jonpas) [inline Scripts/fn_switch.sqf:53]: Why does this need to be a PFH?
https://community.bistudio.com/wiki/Arma_3:_Event_Handlers#AnimStateChanged does this not work?

Comment (TyroneMF) [reply Scripts/fn_switch.sqf:53]: It does not.

**Review by jonpas — APPROVED** [review]

File: Scripts/fn_switchAction.sqf [hunk 5]
```diff
@@ -47,6 +51,16 @@ if (hasInterface) then {
...
 if (isServer) then {
+    // Turn light on.
+    _object animateSource ["SwitchLight", 1, 1];
+
+    // Default switch state, OFF or ON
+    if (_defaultState == 0) then {
+        _object animateSource ["SwitchPosition", -1, 0.5];
+    } else {
+        _object animateSource ["SwitchPosition", 1, 0.5];
+    };
```
Comment (jonpas) [inline Scripts/fn_switchAction.sqf:62]:
```suggestion
    private _switchState = [-1, 1] select _defaultState; // translate default state to switch state (0 = -1, 1 = 1)
    _object animateSource ["SwitchPosition", _switchState, 0.5];
```

**Pull request merged**